# **1. The Euclidean Distance**


In [ ]:
#using euclidean Formula
# Import NumPy Library
import numpy as np
# initializing points in numpy arrays
P1 = np.array((9, 2, 3))
P2 = np.array((1, 4, 7))
# subtracting both the vectors
temp = P1 - P2
# Using Formula
euclid_dist = np.sqrt(np.dot(temp.T, temp))
# printing Euclidean distance
print(euclid_dist)

9.16515138991168


**INFERENCE:**

It find the distance between the points in muti-dimensional vector

Points 1 (a1, b1, c1)

Points 2 (a2, b2, c2)

Euclidean Distance:

sqrt((a1-a2)^2, (b1-b2)^2, (c1-c2)^2)

In [ ]:
# using distance euclidean() method
# Import SciPi Library

from scipy.spatial import distance
# define the points

P1 = (19, 16, 35)
P2 = (11,24, 19)

# print Euclidean distance
print(distance.euclidean(P1,P2))

19.595917942265423


**INFERENCE:**
1. The code imports the distance module from scipy.spatial to compute distances between points.
2. It defines two 3D points: P1 = (19, 16, 35) and P2 = (11, 24, 19).
3. The distance.euclidean() function calculates the Euclidean distance between the two points.
4. The formula used is: sqrt((x2 - x1)^2 + (y2 - y1)^2 + (z2 - z1)^2).
5. The output is approximately 19.60, representing the straight-line distance between the points.

# **2. Manhattans Distance**

In [ ]:
import numpy as np
from scipy.spatial.distance import cityblock

# Example data points
point1 = np.array([1, 2, 3])
point2 = np.array([4, 5, 6])

manhattan_distance = cityblock(point1, point2)
print(f"Manhattan Distance: {manhattan_distance}")

Manhattan Distance: 9


**INFERENCE:**

1. The code imports numpy and the cityblock function from scipy.spatial.distance to compute Manhattan distance.
2. Two 3D points, point1 = [1, 2, 3] and point2 = [4, 5, 6], are defined using numpy arrays.
3. The cityblock() function calculates the Manhattan distance between the points, which is the sum of the absolute differences of their coordinates.
4. The Manhattan distance formula is:
   
   Distance = |x2 - x1| + |y2 - y1| + |z2 - z1|
5. The output printed will be 9, as the sum of the absolute differences is  |4-1| + |5-2| + |6-3| = 3 + 3 + 3 = 9.

# **3. Cosine Similarity**



In [ ]:
import numpy as np
from sklearn.metrics.pairwise import cosine_similarity

# Sample data for items and their features
items_data = {'Item1': [1,2,3], 'Item2': [1, 2, 3], 'Item3': [1,2,3], 'Item4': [1,2,3]}

# Convert the dictionary to a matrix
items_matrix = np.array(list(items_data.values()))
print(items_matrix)

# Calculate cosine similarity
similarity_matrix = cosine_similarity(items_matrix)
print("Similarity Matrix:")
print(similarity_matrix)


[[1 2 3]
 [1 2 3]
 [1 2 3]
 [1 2 3]]
Similarity Matrix:
[[1. 1. 1. 1.]
 [1. 1. 1. 1.]
 [1. 1. 1. 1.]
 [1. 1. 1. 1.]]


**INFERENCE:**

Cosine Similarity:-
Describes the angle between two points
1.   The angle is large when two points are far
2.   The angle is small when two points are near




Point Similarities:

1. Points 1 and 3 (rows 1 and 3) have a very high similarity score of 0.9992, which suggests that these points are nearly identical, despite minor differences in the individual feature values.
2. Points 1 and 4 (rows 1 and 4) also have a very high similarity of 0.9979, suggesting that they are highly similar as well.
3. Points 2 and 4 (rows 2 and 4) have the second-highest similarity at 0.9926, indicating that these points share significant similarities.
4. Points 2 and 3 (rows 2 and 3) have the lowest similarity of 0.9746, although

it is still relatively high. This means that Points 2 and 3 are somewhat less similar compared to other pairs.
Clustering Insight: Based on the similarity values, you might infer that the data points form two primary clusters:

1. Cluster 1: Points 1, 3, and 4 are highly similar to each other and could be considered as part of one cluster.
2. Cluster 2: Point 2 is somewhat less similar to the other points but still has strong similarities with Points 3 and 4. This suggests that while Point 2 might stand apart slightly, it still shares enough similarity to potentially be grouped with the others.

# **4. Jaccard Similarity**

In [ ]:
import numpy as np
from sklearn.metrics import jaccard_score

point1_binary = np.array([1, 1, 0, 0, 1])
point2_binary = np.array([1, 0, 1, 1, 0])

jaccard_similarity = jaccard_score(point1_binary, point2_binary)
print(f"Jaccard Similarity: {jaccard_similarity}")

Jaccard Similarity: 0.2


**INFERENCE:**
1. The code imports numpy and the jaccard_score function from sklearn.metrics to compute the **Jaccard similarity**.
2. Two binary arrays, point1_binary = [1, 1, 0, 0, 1] and point2_binary = [1, 0, 1, 1, 0], represent two sets.
3. The jaccard_score() function calculates the Jaccard similarity, which measures the ratio of the intersection to the union of the two sets.
4. The formula for Jaccard similarity is:

   Jaccard = Intersection of sets / Union of sets
5. The output printed will be 0. as the intersection of the sets is 1 and the union is 5. So, 1/5 = 0.2.

# **5. Minkowski Distance**

In [ ]:
import numpy as np
from scipy.spatial.distance import minkowski

# Example data points
point1 = np.array([1, 2, 3])
point2 = np.array([4, 5, 6])

p = 3
minkowski_distance = minkowski(point1, point2, p)
print(f"Minkowski Distance (p={p}): {minkowski_distance}")

Minkowski Distance (p=3): 4.3267487109222245


**INFERENCE:**
1. The code imports numpy and the minkowski function from scipy.spatial.distance to calculate the Minkowski distance.
2. Two 3D points, point1 = [1, 2, 3] and point2 = [4, 5, 6], are defined using numpy arrays.
3. The variable p = 3 specifies the order of the Minkowski distance, which generalizes other distance metrics
4. The output printed will be the **Minkowski distance** for \( p = 3 \), which is calculated as:

  [ |4-1| ^ 3 + |5-2| ^ 3 + |6-3| ^ 3] ^ (1/3) = (27 + 27 + 27) ^ (1/3) = 81 ^ (1/3) = 4.326